packages/node/src/index.ts

```diff
@@ -528,20 +531,99 @@ export const startDevServer: StartDevServer = async opts => {
     filename: 'package.json',
   });
   const pkg = pathToPkg ? require_(pathToPkg) : {};
+  const tsNodePath = require_.resolve('ts-node');
+  const esmLoader = join(tsNodePath, '..', '..', 'esm.mjs');
+  const cjsLoader = join(tsNodePath, '..', '..', 'register');
+  const isTypescript = ['.ts', '.tsx', '.mts', '.cts'].includes(ext);
+  const maybeTranspile = isTypescript || !['.cjs', '.mjs'].includes(ext);
   const isEsm =
-    entrypoint.endsWith('.mjs') ||
-    (pkg.type === 'module' && entrypoint.endsWith('.js'));
+    ext === '.mjs' ||
+    ext === '.mts' ||
+    (pkg.type === 'module' && ['.js', '.ts', '.tsx'].includes(ext));
 
   const devServerPath = join(__dirname, 'dev-server.js');
+  let nodeOptions = process.env.NODE_OPTIONS;
+  let tsConfig: any = {};
+
+  if (maybeTranspile) {
+    const resolveTypescript = (p: string): string => {
+      try {
+        return require_.resolve('typescript', {
+          paths: [p],
+        });
+      } catch (_) {
+        return '';
+      }
+    };
+
+    const requireTypescript = (p: string): TypescriptModule => require_(p);
+
+    let ts: TypescriptModule | null = null;
+
+    // Use the project's version of Typescript if available and supports `target`
+    let compiler = resolveTypescript(process.cwd());
+    if (compiler) {
+      ts = requireTypescript(compiler);
+    }
+
+    // Otherwise fall back to using the copy that `@vercel/node` uses
+    if (!ts) {
+      compiler = resolveTypescript(join(__dirname, '..'));
+      ts = requireTypescript(compiler);
+    }
+
+    if (pathToTsConfig) {
+      try {
+        tsConfig = ts.readConfigFile(pathToTsConfig, ts.sys.readFile).config;
+      } catch (err) {
+        if (err.code !== 'ENOENT') {
+          console.error(`Error while parsing "${pathToTsConfig}"`);
+          throw err;
+        }
+      }
+    }
+
+    const nodeVersionMajor = Number(process.versions.node.split('.')[0]);
+    fixConfig(tsConfig, nodeVersionMajor);
+
+    // In prod, `.ts` inputs use TypeScript and
+    // `.js` inputs use Babel to convert ESM to CJS.
+    // In dev, both `.ts` and `.js` inputs use ts-node
+    // without Babel so we must enable `allowJs`.
+    tsConfig.compilerOptions.allowJs = true;
+
+    // In prod, we emit outputs to the filesystem.
+    // In dev, we don't emit because we use ts-node.
+    tsConfig.compilerOptions.noEmit = true;
```

Comment:
Was worried that `compilerOptions` might not be set and thus this could throw an error, but then figured out thats happening in `fixConfig`. Maybe we can move these two lines into that function so it all happens in one place?

Reply:
@Ethan-Arrowood `fixConfig()` is used for both prod and dev but these 2 options here are only used for dev